### PR TITLE
server testing with multiple statement, query-timeout and warnings.

### PR DESCRIPTION
**Pull request by princeparmar**
Signed-off-by: pradip parmar <prince.soamedia@gmail.com>

### CODE REVIEW

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"vitess.io/vitess/go/mysql"
+
+	"database/sql"
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+// TestMltistmt check the multiStatements=True and multiStatements=False works properly.
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
// TestMultiStmt checks that multiStatements=True and multiStatements=False work properly.
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+func TestMultistatement(t *testing.T) {
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
func TestMultiStatement(t *testing.T) {
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	ctx := context.Background()
+
+	// connect database with multiStatements=True
+	db := connectDB(t, vtParams, "multiStatements=True", "timeout=90s", "collation=utf8mb4_unicode_ci")
+
+	rows, err := db.QueryContext(ctx, "SELECT 1; SELECT 2; SELECT 3")
+	require.Nilf(t, err, "multiple statement should be executed without error, got %v", err)
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
	require.Nilf(t, err, "multiple statements should be executed without error, got %v", err)
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	_, err = conn.ExecuteFetch("insert into vt_insert_test (id, msg, keyspace_id, data) values(1, 'large blob', 123, 'LLL') /* "+fake.CharactersN(4*1024*1024)+" */", 1, false)
+	require.Nilf(t, err, "insertion error: %v", err)
+
+	qr, err := conn.ExecuteFetch("select * from vt_insert_test where id = 1", 1, false)
+	require.Nilf(t, err, "select error: %v", err)
+	assert.Equal(t, uint64(1), qr.RowsAffected)
+	assert.Equal(t, "BLOB(\"LLL\")", qr.Rows[0][3].String())
+}
+
+// TestInsertLargerThenGrpcLimit insert blob larger then grpc limit and varifies the error.
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
// TestInsertLargerThenGrpcLimit insert blob larger then grpc limit and verify the error.
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	limit, err := strconv.Atoi(grpcLimit)
+	assert.Nilf(t, err, "int parsing error: %v", err)
+
+	// insert data with large blob
+	_, err = conn.ExecuteFetch("insert into vt_insert_test (id, msg, keyspace_id, data) values(2, 'huge blob', 123, '"+fake.CharactersN(limit+1)+"')", 1, false)
+	require.NotNil(t, err, "error expected on insert")
+	assert.Contains(t, err.Error(), "trying to send message larger than max")
+}
+
+// TestTimeout executes sleep(5) with 1s query-timeout, and varifies the error.
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
// TestTimeout executes sleep(5) with query_timeout of 1 second, and verifies the error.
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+func TestTimeout(t *testing.T) {
+	ctx := context.Background()
+
+	conn, err := mysql.Connect(ctx, &vtParams)
+	require.Nilf(t, err, "unable to connect mysql: %v", err)
+	defer conn.Close()
+
+	_, err = conn.ExecuteFetch("SELECT SLEEP(5);", 1, false)
+	require.NotNilf(t, err, "quiry timeout error expected")
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
	require.NotNilf(t, err, "query_timeout error expected")
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	assert.Equal(t, 1317, err.(*mysql.SQLError).Number(), err)
+}
+
+// TestInvalidField tries to fetch invalid column and varifies the error.
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
// TestInvalidField tries to fetch invalid column and verifies the error.
```

File: go/test/endtoend/mysql-server/main_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		}`
+		if err := createConfig(mysqlAuthServerStatic, SQLConfig); err != nil {
+			return 1, err
+		}
+
+		clusterInstance.VtGateExtraArgs = []string{
+			"-vschema_ddl_authorized_users=%",
+			"-mysql_server_query_timeout", "1s",
+			"-mysql_auth_server_impl", "static",
+			"-mysql_server_query_timeout", "1s",
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/main_test.go]: This line is repeated, delete one of them.

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	// validate with 0 warnings
+	_, err = conn.ExecuteFetch("SELECT 1 from vt_insert_test limit 1", 1, false)
+	require.Nilf(t, err, "select error: %v", err)
+
+	qr, err = conn.ExecuteFetch("SHOW WARNINGS;", 1, false)
+	require.Nilf(t, err, "SHOW WARNINGS; execution failed: %v", err)
+	assert.Equalf(t, uint64(0), qr.RowsAffected, "0 warning expected, got %v ", qr.RowsAffected)
+}
+
+// TestSelectWithUnauthorizedUser varifies that unauthorized user
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
// TestSelectWithUnauthorizedUser verifies that an unauthorized user
```

File: go/test/endtoend/mysql-server/mysql_server_test.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+// is not able to read from the table.
+func TestSelectWithUnauthorizedUser(t *testing.T) {
+	ctx := context.Background()
+
+	tmpVtParam := vtParams
+	tmpVtParam.Uname = "testuser2"
+	tmpVtParam.Pass = "testpassword2"
+
+	conn, err := mysql.Connect(ctx, &tmpVtParam)
+	require.Nilf(t, err, "unable to connect mysql: %v", err)
```
Comment (deepthi) [inline go/test/endtoend/mysql-server/mysql_server_test.go]:
```suggestion
	require.Nilf(t, err, "unable to connect to mysql: %v", err)
```

File: test/config.json [hunk 11]
```diff
@@ -442,6 +433,15 @@
 			"RetryMax": 0,
 			"Tags": []
 		},
+		"mysql_server": {
+			"File": "mysql_server_test.go",
+			"Args": ["vitess.io/vitess/go/test/endtoend/mysql-server"],
```
Comment (deepthi) [inline test/config.json]: rename the directory to mysqlserver

**Comment by morgo** [comment]
> The concern is that we are introducing new dependencies. The fear I have is that this may make the project more fragile.
> 
> On the other hand, this test proves that vitess works with other drivers or ORMs, which is a good thing.
> 
> I want to see if others have any thoughts about this.

I will chime in with mine. I see this as a choice between stability or compatibility. I prefer compatibility in this specific case.

i.e. if vttablet used go-sql-driver/mysql, I don't think we would have had to work on https://github.com/vitessio/vitess/pull/5683 ; I assume in future we will need to add `caching_sha2_password` https://github.com/vitessio/vitess/issues/5399 whereas this is already supported in  go-sql-driver/mysql.

**Comment by deepthi** [comment]
Can you rebase or merge and resolve the conflicts?

**Comment by ajeetj** [comment]
@deepthi  Please use Squash & Merge for this PR.

**Pull request merged**